components/core/src/clp_s/JsonParser.cpp

```diff
@@ -557,6 +621,292 @@ int32_t JsonParser::add_metadata_field(std::string_view const field_name, NodeTy
     return m_archive_writer->add_node(metadata_subtree_id, type, field_name);
 }
 
+auto JsonParser::get_archive_node_type(
+        clp::ffi::SchemaTree::Node::Type ir_node_type,
+        bool node_has_value,
+        std::optional<clp::ffi::Value> const& node_value
+) -> NodeType {
+    // figure out what type the node is in archive node type
+    NodeType archive_node_type = NodeType::Unknown;
+    switch (ir_node_type) {
+        case clp::ffi::SchemaTree::Node::Type::Int:
+            archive_node_type = NodeType::Integer;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Float:
+            archive_node_type = NodeType::Float;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Bool:
+            archive_node_type = NodeType::Boolean;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::UnstructuredArray:
+            archive_node_type = NodeType::UnstructuredArray;
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Str:
+            if (node_value && node_value->is<std::string>()) {
+                archive_node_type = NodeType::VarString;
+            } else {
+                archive_node_type = NodeType::ClpString;
+            }
+            break;
+        case clp::ffi::SchemaTree::Node::Type::Obj:
+            if (node_has_value) {
+                if (node_value->is_null()) {
+                    archive_node_type = NodeType::NullValue;
+                } else {
+                    archive_node_type = NodeType::Object;
+                }
+            } else {
+                archive_node_type = NodeType::Object;
+            }
+            break;
+        default:
+            break;
+    }
+    return archive_node_type;
+}
+
+auto JsonParser::get_archive_node_id(
+        std::unordered_map<uint32_t, std::vector<std::pair<NodeType, int32_t>>>&
+                ir_node_to_archive_node_unordered_map,
+        uint32_t ir_node_id,
+        NodeType archive_node_type,
+        clp::ffi::SchemaTree const& ir_tree
+) -> int {
+    auto unordered_map_location = ir_node_to_archive_node_unordered_map.find(ir_node_id);
+    if (ir_node_to_archive_node_unordered_map.end() != unordered_map_location) {
+        auto translation_vector = unordered_map_location->second;
+        for (auto& i : translation_vector) {
+            if (i.first == archive_node_type) {
+                return i.second;
+            }
+        }
+    }
+
+    auto const& curr_node = ir_tree.get_node(ir_node_id);
+    int32_t parent_node_id{-1};
+    auto parent_of_curr_node_id = curr_node.get_parent_id();
+    if (parent_of_curr_node_id.has_value()) {
+        parent_node_id = get_archive_node_id(
```

Comment:
Will leave it up to you whether to do this change right now since it's maybe a bit of a more complicated rewrite, but we should at least keep it in mind for a follow-up PR.